Meta: Add templates/checklists for new, Accepted/Rejected & Final PEPs

CAM-Gerlach opened the pull request:
As discussed in #2937 , this adds new templates for key PEP state transitions with concise checklists for each of the steps needed, to help both authors and reviewers. This relies on a new, experimental/alpha GitHub feature (enabled only on our repo) to work, so hopefully this is doing it right, but the only way to know is to merge this.

If it works, we can follow this up with a couple more templates for the other two common cases (updating a draft PEP and changing an old PEP), which don't need a checklist or more text, but can contain additional reminders tailored to that situation and be an opportunity to inform users what types of PRs are welcome and redirect/point them to more resources.

AA-Turner commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+* [ ] ``PEP``, ``Title``, ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
+* [ ] ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` filled out if appropriate
+* [ ] ``Sponsor`` listed and approves if not authored by a core dev or PEP editor
+* [ ] PEP builds with no warnings, CI checks pass and content displays as intended in the rendered preview
```
This is hard to check ex-ante, given the preview only starts when the PR is opened.

CAM-Gerlach replied:
Well, they can always build it locally, check the preview and run the linting before they submit their final PR—that's fully equivalent to the CI. And as mentioned completion any item of the checklist prior to PR submission is entirely optional (though suggested).

CAM-Gerlach replied:
I revisied this to change "CI checks" to "pre-commit checks" and "rendered preview" to "rendered HTML", to make it apply equally to local and CI builds.

CAM-Gerlach replied:
Marking this as addressed; feel free to re-open or add another comment with any further feedback.

AA-Turner commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+* [ ] ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` filled out if appropriate
+* [ ] ``Sponsor`` listed and approves if not authored by a core dev or PEP editor
+* [ ] PEP builds with no warnings, CI checks pass and content displays as intended in the rendered preview
+* [ ] PEP number assigned by the PEP editors and filename & ``PEP`` header updated accordingly
```
Again I think this needs to be clearer that assignment only happens after the PR is opened (e.g. a section *for PEP Editor use*).

CAM-Gerlach replied:
I revised the language to reflect that the PEP should be given the next available number, rather than necessarily requiring assignment by the PEP editors. I also was able to consolidate the filename item into this one, and fixed up the grammar/syntax a bit.

CAM-Gerlach replied:
Marking this as addressed; feel free to re-open or add another comment with any further feedback.

encukou commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
+* [ ] Title clearly, accurately and concisely describes the content
+* [ ] Prose is intelligible and generally free of basic grammar and syntax errors, at least to the degree they significantly detract from understanding
+* [ ] Code is in code blocks, well-formatted (PEP 7/PEP 8) and has the right [lexer name](https://pygments.org/docs/lexers/) if non-Python (or ``text``)
```
This could use a Python-specific section in the devguide. The Pygments docs are too broad.

CAM-Gerlach replied:
We already [discuss the specific usage of code blocks and the lexer name](https://peps.python.org/pep-0012/#literal-blocks) in PEP 12; I'll go ahead and link that there—good idea.

```suggestion
* [ ] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer name](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
```

hugovk commented on .github/PULL_REQUEST_TEMPLATE/accept-reject-pep.md:
```diff
@@ -0,0 +1,50 @@
+* [ ] ``Status`` changed to ``Accepted``/``Rejected``
+* [ ] ``Resolution`` link points directly to SC/PEP Delegate official acceptance/rejected post
+* [ ] Acceptance/rejection notice added, if the SC/PEP delegate had any major conditions or critical commentary (and any changes made accordingly)
+* [ ] ``Discussions-To``, ``Post-History` and ``Python-Version`` complete and up to date
```
Missing backtick here (although single backticks are more usual for Markdown (unlike reST))

```suggestion
* [ ] ``Discussions-To``, ``Post-History`` and ``Python-Version`` complete and up to date
```

CAM-Gerlach replied:
Implemented in Petr's suggestion, thanks!

encukou commented on .github/PULL_REQUEST_TEMPLATE/accept-reject-pep.md:
```diff
@@ -0,0 +1,50 @@
+
+
+# Accept/Reject PEP
+
+<!--
+
+Link to the official acceptance/rejection message
+by the Steering Council/PEP Delegate on the PEP's discussion thread.
+
+-->
+
+
+
+## Checklist
+
+<!--
+
+This lists the key steps needed to mark a PEP as Accepted/Rejected.
```
The link is in the proposed PEP text, so I don't think it *needs* to be in the PR message.

```suggestion

<!--

This checklist contains the key changes to mark a PEP as Accepted/Rejected.
```

encukou commented on .github/PULL_REQUEST_TEMPLATE/mark-pep-final.md:
```diff
@@ -0,0 +1,48 @@
+
+
+# Mark PEP Final
+
+<!--
+
+State why the PEP should be marked Final,
+linking any relevant discussions, issues or PRs.
+
+-->
+
+
+
+## Checklist
+
+<!--
+
+This lists the key steps needed to mark a PEP as Final.
```
```suggestion

<!--

This checklist contains the key changes needed to mark a PEP as Final.
```

encukou commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+
+
+# New PEP
+
+<!--
+
+Briefly describe your new PEP here, and anything else we should know.
+Be sure to link to any previous discussions or issues motivating/drafting it.
+
+-->
+
+
+
+## Checklist
+
+<!--
+
+You should use the following as a guide when double-checking your PEP,
+and you can help complete some of it yourself if you like
+by ticking any boxes you're sure about, like this: [x]
+
+If you're unsure about anything, just leave it blank and we'll take a look.
+
+-->
+
```
```suggestion

<!--

Resist the temptation to describe your PEP here.
Your PEP's Abstract and Post-History should speak for itself.

You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.
-->

```

CAM-Gerlach replied:
```suggestion

<!--

You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

-->

```

The above, minus the extra "resist the temptation" text—this hasn't ever really been an actual problem in the past, and the more structured nature of this template should likely further dissuade if it is. Additionally, a brief meta-description with some additional contextually-relevant details doesn't exactly _hurt_, if authors feel the need to provide it.

encukou reviewed (commented):
Here are some thoughts and attempts to simplify things.

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/mark-pep-final.md:
```diff
@@ -0,0 +1,48 @@
+-->
+
+
+* [ ] Specification has been implemented in a released stable CPython version (or third-party tools, for packaging/typing topic PEPs)
+* [ ] ``Status`` changed to ``Final``
+* [ ] Canonical docs/spec moved to an appropriate place, and linked with a ``canonical-doc`` directive (or ``pypa-spec``, for packaging PEPs)
+* [ ] PEP headers and content updated with any substantial changes during the implementation phase
```
```suggestion
* [ ] Final implementation has been merged (including tests and documentation)
* [ ] PEP matches the final implementation
* [ ] Any substantial changes since the accepted version approved by the SC/PEP delegate
* [ ] ``Status`` changed to ``Final``
* [ ] Canonical docs/spec linked with a ``canonical-doc`` directive (or ``pypa-spec``, for packaging PEPs)
```

I've adopted a revised form of your suggestion above, with some further trimming down

> IMO the canonical-spec point should link to a devguide page that documents the directives.

We certainly should document it somewhere (in a separate issue/PR). Right now it is PEP 12 that documents all the rest of the syntax, directives and formatting for PEPs, so presumably it should go there for now, rather than the devguide which doesn't currently contain any other PEP-specific content. (As mentioned previously a potential future move to the devguide could make sense, but this should be discussed, decided upon and implemented for all the similar information as opposed to piecemeal, resulting in a fragmentary experience for authors and PEP editors).

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+---
+name: New PEP
+about: Submit a new draft PEP for consideration
```
```suggestion
name: New Standards Track PEP
about: Submit a new draft Standards Track PEP for consideration
```

Adopted the above, using consistent Title Case as in the canonical Type header as opposed to a rather unusual (in English) mid-sentence capitalization of only Standard

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+---
+name: New PEP
+about: Submit a new draft PEP for consideration
```
Good point, though a lot of it is still fully applicable—presumably, we should make another (shorter/more general) checklist for non-standards-track PEPs with the relevant elements. I can take a crack at it.

CAM-Gerlach replied:
After experimenting with several options—including separate Process/Information PEP checklist, and separate Process and Information checklists—I ended up re-organizing the existing new PEP checklist into a section that applied to all PEPs, and a section specific to Standards Track PEPs (that authors were instructed to remove if not applicable). I also caught a few other places (e.g. in the two other templates) where things only apply to Standards Track PEPs, and clarified them accordingly, as well as some other tweaks.

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+
+### Content requirements
+
+* [ ] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) and general agreement that a PEP is appropriate
```
That's a good practice and should be at least strongly encouraged if not required, but I'm not sure whether it is critical to explicitly state such in this checklist item (as opposed to, e.g., the PEP 1/12 mention of Post-History), given Post-History is mentioned below.

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+* [ ] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) and general agreement that a PEP is appropriate
+* [ ] Content is reasonably sound, complete and makes at least basic technical sense
+* [ ] Abstract and Copyright sections completed, with the required wording in the latter
+* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
```
Yeah, I'd certainly find that useful as both a PEP author and reviewer, I was just worried that it would make it too long, but I've gone and and implemented it in another comment

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+
+* [ ] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) and general agreement that a PEP is appropriate
+* [ ] Content is reasonably sound, complete and makes at least basic technical sense
+* [ ] Abstract and Copyright sections completed, with the required wording in the latter
+* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
```
```suggestion
* [ ] [Required and suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [ ] Abstract (required; first section)~
    * [ ] Motivation
    * [ ] Rationale
    * [ ] Specification
    * [ ] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [ ] Reference Implementation
    * [ ] Rejected Ideas
    * [ ] Open Issues
    * [ ] Copyright (required with exact wording; last section)
```

As [suggested by @encukou above](https://github.com/python/peps/pull/2956/files#r1065559112)

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+* [ ] Content is reasonably sound, complete and makes at least basic technical sense
+* [ ] Abstract and Copyright sections completed, with the required wording in the latter
+* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
+* [ ] Title clearly, accurately and concisely describes the content
+* [ ] Prose is intelligible and generally free of basic grammar and syntax errors, at least to the degree they significantly detract from understanding
```
```suggestion
* [ ] Title clearly, accurately and concisely describes the content in 79 characters or less
```

IMO, it's worth mentioning this one explicitly, since it's a header field, has a soft length limit and needs special attention by authors and reviewers (indeed, on the most recent new PEP I reviewed, PEP 702, had a title in need of much adjustment to meet this standard, and this checklist was specifically useful as a reminder of such).

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,72 @@
+
+* [ ] Filename in the correct format (``pep-NNNN.rst``)
+* [ ] ``PEP``, ``Title``, ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
+* [ ] ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` filled out if appropriate
+* [ ] ``Sponsor`` listed and approves if not authored by a core dev or PEP editor
```
```suggestion
* [ ] ``Sponsor``, ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` completed if appropriate
```

Fixed order to match headers, added missing "and" and fixed (original) wording

CAM-Gerlach commented on .github/PULL_REQUEST_TEMPLATE/accept-reject-pep.md:
```diff
@@ -0,0 +1,50 @@
+* [ ] SC/PEP Delegate has formally accepted/rejected the PEP and posted to the ``Discussions-To`` thread
+* [ ] ``Status`` changed to ``Accepted``/``Rejected``
+* [ ] ``Resolution`` link points directly to SC/PEP Delegate official acceptance/rejected post
+* [ ] Acceptance/rejection notice added, if the SC/PEP delegate had any major conditions or critical commentary (and any changes made accordingly)
+* [ ] ``Discussions-To``, ``Post-History` and ``Python-Version`` complete and up to date
```
Added missing backtick from Hugo's suggestion

CAM-Gerlach commented:
Sorry for the delay; I started to go over this several days ago but accidentally marked the notification as "done" instead of "unread", and so it slipped through the cracks for several days. In addition to the suggested revisions above (with some adaptations), and a couple inspired by them, I pushed another commit re-organizing and revising the new PEP checklist more thoroughly to account for Standards Track vs. other types of PEPs, as well as tweaking the other checklists to clarify this too

I also attempted to address both of Adam/Petr's two points above, as well as fixed/revised a few other things I noticed. Looking forward to your further feedback!

encukou commented:
Thanks! Tooks good to merge now!

I still think the two comments and intervening heading is superfluous though. Shouldn't “Make sure to include the PEP number in the pull request title” be a checklist item too?

CAM-Gerlach commented:
Thanks, that makes sense. I've gone ahead and removed the first comment block entirely on all the templates (except for other), added the relevant steps (PR title, reading PEP 1/12) to new or existing checklist items, and also slimmed down the second comment block on each.

hugovk reviewed (approved):
Thanks!

CAM-Gerlach commented:
Based on @encukou 's previous points above and Brett's comment on #2966 , I renamed the existing "Other" template to be for "Update/modify PEP", greatly slimmed down the HTML comment and moved the "PR title" to be a single-line checkbox rather than a multi-line comment, which makes it take up less space but be more easily visible. Furthermore, I added a set of checkboxes for the types of changes that are welcomed: to a draft PEP, to an accepted/final PEP with SC approval, or resolving a purely technical or editorial defect. Overall, this makes it much more visible and useful, while actually taking up several lines less than the original.

I also added back a much more minimal "Other" template for infra, meta, etc. PRs with just a two-line reminder comment to check the contributing guide and make sure this repo is the right place for the change, which makes it a fraction the length of the original and cuts out material not relevant to that type of PR (PEP number in title, etc).

Finally, I was able to trim a few more lines from the comments in the other templates, without any substantive affect on the content.

The new `Update/modify PEP` template renders like this:

----

<!--
**Please** read our Contributing Guidelines (CONTRIBUTING.rst)
to make sure this repo is the right place for your proposed change. Thanks!
-->

* Change is either:
    * [ ] To a Draft PEP
    * [ ] To an Accepted or Final PEP, with Steering Council approval
    * [ ] Resolving a purely technical or editorial issue
* [ ] PR title prefixed with PEP number (e.g. ``PEP 123: Summary of changes``)

encukou commented on .github/PULL_REQUEST_TEMPLATE/modify-pep.md:
```diff
@@ -0,0 +1,15 @@
+* Change is either:
+    * [ ] To a Draft PEP
+    * [ ] To an Accepted or Final PEP, with Steering Council approval
+    * [ ] Resolving a purely technical or editorial issue
```
I'd read that as resolving technical issues in the proposal. This should only be about issues in how the proposal is described :)
```suggestion
    * [ ] Fixing an editorial issue (e.g. typo)
```

CAM-Gerlach replied:
Yeah, I went back and forth on that, but yeah you're right—it's best to avoid misinterpretation, and generally speaking we're going to be the ones fixing technical issues of the sort I meant (warnings, links, Sphinx syntax, headers, etc.) which arguably fall under the general umbrella of editorial issues anyway.

I do suggest if we mention examples, we do so more broadly than jsut typos, or we may encourage users spending their time going looking for them, like we've had a number of PRs of in the past (perhaps even running automated tooling) as opposed to much more useful things like the stuff I list in https://github.com/python/peps/issues/10#issuecomment-1382420957 (and will list and describe in more detail in the forthcoming meta-issue mentioned there).

So, e.g.:

```suggestion
    * [ ] To fix an editorial issue (markup, typo, link, header, etc)
```

CAM-Gerlach replied:
Going ahead with this suggestion for now

encukou commented on .github/PULL_REQUEST_TEMPLATE/other.md:
```diff
@@ -0,0 +1,9 @@
+---
+name: Other
+about: Infra or meta change not falling in another category
```
I don't think it's wise to limit the “other” category.
```suggestion
about: Any change not falling in another category
```

CAM-Gerlach replied:
Yeah perhaps not, I didn't have it limited originally, but my concern was that users (especially those who would be least familiar with the things mentioned in the checklist) would be more likely to immediately go for a generic "Other" rather than "Changing a PEP" if it wasn't explicitly qualified like this, and on this repo, any valid change that isn't "Changing a PEP" is, by definition, a change to the infrastructure (rendering system, workflows, makefile, config, etc) and/or a meta change, so I'm not sure what (acceptable) PR could plausibly not fall under any of those categories.

hugovk replied:
Marking as resolved to be able to merge. We can adjust this later if needed.

CAM-Gerlach commented:
As I noticed Hugo using it on python/core-workflow#491, I've added a `title:` field to the appropriate templates, which pre-fills the title following the given format, which should help a lot with following the standard title format. Of course, it _might_ not actually work yet, but same with all the other fields, only one way to find out...

hugovk commented on .github/PULL_REQUEST_TEMPLATE/accept-reject-pep.md:
```diff
@@ -0,0 +1,21 @@
+---
+name: Accept/Reject PEP
+about: Mark a draft Standards Track PEP as accepted or rejected
+title: PEP NNN: Mark as Accepted
```
Viewing this file in the GitHub UI:

<img width="797" alt="image" src="https://user-images.githubusercontent.com/1324225/214009924-f23af56f-2d33-4bd7-94e2-4115392c3f10.png">

https://github.com/CAM-Gerlach/peps/blob/infra-pr-checklists/.github/PULL_REQUEST_TEMPLATE/accept-reject-pep.md

Where line 3 refers to the `title:` line. It doesn't like the repeated colon.

Editing in the UI and adding quotes, shows no warning in the preview:

```suggestion
title: "PEP NNN: Mark as Accepted"
```

CAM-Gerlach replied:
Oops, I was wondering about that and am usually pretty conservative quoting YAML, but in this case I forgot to check. Thanks!

hugovk commented on .github/PULL_REQUEST_TEMPLATE/mark-pep-final.md:
```diff
@@ -0,0 +1,20 @@
+---
+name: Mark PEP Final
+about: Mark an Accepted Standards Track PEP as Final
+title: PEP NNN: Mark Final
```
```suggestion
title: "PEP NNN: Mark Final"
```

hugovk commented on .github/PULL_REQUEST_TEMPLATE/modify-pep.md:
```diff
@@ -0,0 +1,16 @@
+---
+name: Update/modify PEP
+about: Add to or change an existing PEP
+title: PEP NNN: Description of change
```
```suggestion
title: "PEP NNN: Description of change"
```

hugovk commented on .github/PULL_REQUEST_TEMPLATE/new-pep.md:
```diff
@@ -0,0 +1,50 @@
+---
+name: New PEP
+about: Submit a new draft PEP
+title: PEP NNN: Title of PEP
```
```suggestion
title: "PEP NNN: Title of PEP"
```

hugovk reviewed (approved):
Thanks! I think we're good to merge (and iterate if need be)?

CAM-Gerlach commented:
Due to it specifically coming up on the past two new PEPs, and the most recent new PEP having an issue where the listed sponsor didn't actually approve being listed, I've added back the explicit checklist item for new PEPs to either have a core dev/PEP editor as an author, or formally confirmed as a sponsor.

hugovk commented:
Think we're good to merge this, we can always adjust later as needed.

Thank you!

CAM-Gerlach commented:
Okay, I'm glad I was around to check if it actually worked, as I had no idea if it would and was hoping to be around when this was merged—I had very limited info from GitHub, since it's all super experimental.

Indeed, it does work in a basic sense, but it looks like the current UI just shows the file names, and the data block at the top of the templates is not used currently like it is for the issues:

![image](https://user-images.githubusercontent.com/17051931/216155368-2319c905-d3b6-4c18-8908-907540df3659.png)

With that in mind, we presumably should just remove it for now and rejigger the file names so they're more informative, at least as a quick stopgap.

hugovk commented:
You also don't get the selector when doing "New pull request" from `https://github.com/username/peps/branches`, but I guess that's a less common path than clicking the link from https://github.com/python/peps for a recently pushed branch.

CAM-Gerlach commented:
Hmm, good point—I didn't even realize you could do that, I always just use the recently pushed branch link or set it up manually. That's possibly because the "compare" UI (which the template selector is on) never triggers when going that route, and it instead jumps straight to the PR creation. We can mention that to GitHub.

I've opened #2998 to fix the issues mentioned here; hopefully we can get that merged quickly to avoid any impact on contributors.
